posts/2024-05-06-jakarta-nosql-in-action-meet-jnopo-game.adoc

```diff
@@ -0,0 +1,828 @@
+                    PlayerInfo.of(gameOver.playerBInfo()),
+                    gameOver.winnerInfo().map(PlayerInfo::of).orElse(PlayerInfo.NOBODY),
+                    gameOver.winnerInfo().map(PlayerInfo::of).orElse(PlayerInfo.NOBODY),
+                    gameOver.isTied()
+            );
+
+            playoffs.add(gameMatch);
+    }
+
+}
+----
+
+Now, let's get started the  https://openliberty.io/[Open Liberty] runtime to figure out if the game match results are being stored on the MongoDB:
+
+* On Linux/Mac machines, perform the following command:
++
+[source, bash]
+----
+./mvnw liberty:dev
+----
+
+* On Windows machines, perform the following command:
++
+[source, powershell]
+----
+mvnw.cmd liberty:dev
+----
+
+After some game matches, we can confirm the persistence of the game matches into the database by looking at Mongo Express running on http://localhost:8081 :
+
+[.img_border_light]
+image::/img/blog/mongo-express-01.png[Mongo Express 01, align="center" width=80%]
+
+[.img_border_light]
+image::/img/blog/mongo-express-02.png[Mongo Express 02, align="center" width=80%]
+
+[.img_border_light]
+image::/img/blog/mongo-express-03.png[Mongo Express 03, align="center" width=80%]
+
+[.img_border_light]
+image::/img/blog/mongo-express-04.png[Mongo Express 04, align="center" width=80%]
+
+Or, if you're a command-line practitioner, you can check the data by using the `mongosh` cli:
+
+[source, bash]
+----
+docker exec -it finish-mongo-1 mongosh -u root -p example --authenticationDatabase admin jnopo
+----
+
+Once connected, feel free to perform mongo commands:
+
+* Getting the number of stored documents:
++
+[source, bash]
+----
+jnopo> db.GameMatch.countDocuments()
+3
+----
+
+* Getting the stored documents list:
++
+[source, bash]
+----
+jnopo> db.GameMatch.find()
+[
+  {
+    _id: '2f42cf9e-f99c-44e5-bda8-7707da2579a3',
+    tied: true,
+    winner: { name: '', movement: '' },
+    playerA: { name: 'Max', movement: 'ROCK' },
+    playerB: { name: 'Ghost', movement: 'ROCK' },
+    loser: { name: '', movement: '' }
+  },
+  {
+    _id: 'c425815c-2aaf-4caa-a975-8e8089f0a0ce',
+    tied: false,
+    winner: { name: 'Max', movement: 'PAPER' },
+    playerA: { name: 'Ghost', movement: 'ROCK' },
+    playerB: { name: 'Max', movement: 'PAPER' },
+    loser: { name: 'Max', movement: 'PAPER' }
+  },
+  {
+    _id: 'bb7e7cd5-f8fe-4db1-9e90-44ecd433b4de',
+    tied: false,
+    winner: { name: 'Ghost', movement: 'ROCK' },
+    playerA: { name: 'Ghost', movement: 'ROCK' },
+    playerB: { name: 'Max', movement: 'SCISSORS' },
+    loser: { name: 'Ghost', movement: 'ROCK' }
+  }
+]
+----
+
+Great! *JNopo* is storing the game match results as expected! It's time to create the winner ranking and then expose it though a restful endpoint http://localhost:9080/jakarta-nosql-game/api/playoffs/ranking .
+
+==== Exposing the Winner Ranking
+
+Collecting and storing the game matches results is a great step to make us able to implement the winner ranking feature.
```

Comment:
```suggestion
Collecting and storing the game match results is a great step to enable us able to implement the winner ranking feature. Let's implement the `Ranking` class to represent any ranking on the *JNopo* application. The `Ranking` creation requires data from the `Playoffs` component. In order to avoid having to create a new layer to keep the logic of the ranking creation, we're going to use a simple approach: a simple static factory method on the `Ranking` class itself.
```